networking/eth2/src/main/java/tech/pegasys/artemis/networking/eth2/gossip/topics/validation/AttestationValidator.java

```diff
@@ -0,0 +1,212 @@
+    if (isAfterPropagationSlotRange(currentTimeMillis, attestation)) {
+      return INVALID;
+    }
+    if (isBeforeMinimumBroadcastTime(attestation, currentTimeMillis) < 0) {
+      return SAVED_FOR_FUTURE;
+    }
+
+    // The block being voted for (attestation.data.beacon_block_root) passes validation.
+    // It must pass validation to be in the store.
+    // If it's not in the store, it may not have been processed yet so save for future.
+    final Optional<BeaconState> maybeState =
+        recentChainData.getBlockState(attestation.getData().getBeacon_block_root());
+    if (maybeState.isEmpty()) {
+      return SAVED_FOR_FUTURE;
+    }
+
+    final BeaconState state = maybeState.get();
+
+    // The signature of attestation is valid.
+    final IndexedAttestation indexedAttestation = get_indexed_attestation(state, attestation);
+    if (!is_valid_indexed_attestation(state, indexedAttestation)) {
+      return INVALID;
+    }
+    return VALID;
+  }
+
+  private ValidatorAndSlot getValidatorAndSlot(final Attestation attestation) {
+    return new ValidatorAndSlot(
+        attestation.getData().getSlot(),
+        attestation.getData().getIndex(),
+        attestation.getAggregation_bits().streamAllSetBits().findFirst().orElseThrow());
+  }
+
+  private int isBeforeMinimumBroadcastTime(
+      final Attestation attestation, final UnsignedLong currentTimeMillis) {
+    final UnsignedLong minimumBroadcastTimeMillis =
+        minimumBroadcastTimeMillis(attestation.getData().getSlot());
+    return currentTimeMillis.compareTo(minimumBroadcastTimeMillis);
+  }
+
+  private boolean isAfterPropagationSlotRange(
+      final UnsignedLong currentTimeMillis, final Attestation aggregate) {
+    final UnsignedLong aggregateSlot = aggregate.getData().getSlot();
+    return maximumBroadcastTimeMillis(aggregateSlot).compareTo(currentTimeMillis) < 0;
+  }
+
+  private UnsignedLong secondsToMillis(final UnsignedLong seconds) {
+    return seconds.times(MILLIS_PER_SECOND);
+  }
+
+  private UnsignedLong minimumBroadcastTimeMillis(final UnsignedLong aggregateSlot) {
+    final UnsignedLong lastAllowedTime =
+        recentChainData
+            .getGenesisTime()
+            .plus(aggregateSlot.times(UnsignedLong.valueOf(SECONDS_PER_SLOT)));
+    final UnsignedLong lastAllowedTimeMillis = secondsToMillis(lastAllowedTime);
+    return lastAllowedTimeMillis.compareTo(MAXIMUM_GOSSIP_CLOCK_DISPARITY) >= 0
+        ? lastAllowedTimeMillis.minus(MAXIMUM_GOSSIP_CLOCK_DISPARITY)
+        : ZERO;
+  }
+
+  private UnsignedLong maximumBroadcastTimeMillis(final UnsignedLong aggregateSlot) {
+    final UnsignedLong lastAllowedSlot = aggregateSlot.plus(ATTESTATION_PROPAGATION_SLOT_RANGE);
+    // The last allowed time is the end of the lastAllowedSlot.
+    final UnsignedLong lastAllowedTime =
+        recentChainData
+            .getGenesisTime()
+            .plus(lastAllowedSlot.plus(ONE).times(UnsignedLong.valueOf(SECONDS_PER_SLOT)));
```

Comment:
Because you're allowed to send the attestation until the end of that slot.  `lastAllowedSlot * SECONDS_PER_SLOT` will give you the start of the slot and we want the end.